drivers/pca9685/pca9685.c

```diff
@@ -0,0 +1,369 @@
+
+    /* set Auto-Increment flag */
+    uint8_t byte = 0;
+    _set_reg_bit(&byte, PCA9685_MODE2_INVERT, dev->params.inv);
+    _set_reg_bit(&byte, PCA9685_MODE2_OUTDRV, dev->params.out_drv);
+    _set_reg_bit(&byte, PCA9685_MODE2_OUTNE, dev->params.out_ne);
+    EXEC_RET(_write(dev, PCA9685_REG_MODE2, &byte, 1));
+
+    /* set Sleep mode, Auto-Increment, Restart, All call and External Clock */
+    byte = 0;
+    _set_reg_bit(&byte, PCA9685_MODE1_AI, 1);
+    _set_reg_bit(&byte, PCA9685_MODE1_SLEEP, 1);
+    _set_reg_bit(&byte, PCA9685_MODE1_RESTART, 1);
+    _set_reg_bit(&byte, PCA9685_MODE1_ALLCALL, 1);
+    EXEC_RET(_write(dev, PCA9685_REG_MODE1, &byte, 1));
+
+    /* must be done only in sleep mode */
+    _set_reg_bit(&byte, PCA9685_MODE1_EXTCLK, dev->params.ext_freq ? 1 : 0);
+    EXEC_RET(_write(dev, PCA9685_REG_MODE1, &byte, 1));
+
+    return PCA9685_OK;
+}
+
+static int _read(const pca9685_t *dev, uint8_t reg, uint8_t *data, uint32_t len)
+{
+    DEBUG_DEV("reg=%02x data=%p len=%"PRIu32"", dev, reg, data, len);
+
+    /* acquire the I2C device */
+    if (i2c_acquire(dev->params.i2c_dev)) {
+        DEBUG_DEV("could not aquire I2C bus", dev);
+        return -PCA9685_ERROR_I2C;
+    }
+
+    if (i2c_read_regs(dev->params.i2c_dev,
+                      dev->params.i2c_addr, reg, data, len, 0) != 0) {
+        i2c_release(dev->params.i2c_dev);
+        return -PCA9685_ERROR_I2C;
+    }
+
+    /* release the I2C device */
+    i2c_release(dev->params.i2c_dev);
+
+    return PCA9685_OK;
+}
+
+static int _write(const pca9685_t *dev, uint8_t reg, const uint8_t *data, uint32_t len)
+{
+    DEBUG_DEV("reg=%02x data=%p len=%"PRIu32"", dev, reg, data, len);
+
+    if (i2c_acquire(dev->params.i2c_dev)) {
+        DEBUG_DEV("could not aquire I2C bus", dev);
+        return -PCA9685_ERROR_I2C;
+    }
+
+    if (i2c_write_regs(dev->params.i2c_dev,
+                       dev->params.i2c_addr, reg, data, len, 0) != 0) {
+        i2c_release(dev->params.i2c_dev);
+        return -PCA9685_ERROR_I2C;
+    }
+
+    /* release the I2C device */
+    i2c_release(dev->params.i2c_dev);
+
+    return PCA9685_OK;
+}
+
+inline static int _write_word(const pca9685_t *dev, uint8_t reg, uint16_t data)
+{
+    uint8_t bytes[2] = { data & 0xff, (data >> 8) & 0xff };
+    return _write (dev, reg, bytes, 2);
+}
+
+static int _update(const pca9685_t *dev, uint8_t reg, uint8_t mask, uint8_t data)
+{
+    uint8_t byte;
+
+    /* read current register value */
+    EXEC_RET(_read(dev, reg, &byte, 1));
+
+    /* set masked bits to the given value  */
+    _set_reg_bit(&byte, mask, data);
+
+    /* write back new register value */
+    EXEC_RET(_write(dev, reg, &byte, 1));
+
+    return PCA9685_OK;
+}
+
+static void _set_reg_bit(uint8_t *byte, uint8_t mask, uint8_t bit)
+{
+    uint8_t shift = 0;
+    while (!((mask >> shift) & 0x01)) {
+        shift++;
+    }
+    *byte = ((*byte & ~mask) | ((bit << shift) & mask));
```

Comment:
Turns out the compiler will [optimize this all away](https://godbolt.org/z/t38n6f) - neat!